debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlSnapshotChangeEventSource.java

```diff
@@ -333,38 +346,77 @@ protected void readTableStructure(ChangeEventSourceContext sourceContext,
+                }
 
-            LOGGER.info("Reading structure of database '{}'", database);
-            addSchemaEvent(snapshotContext, database, "DROP DATABASE IF EXISTS " + quote(database));
-            final StringBuilder createDatabaseDddl = new StringBuilder("CREATE DATABASE " + quote(database));
-            final DatabaseLocales defaultDatabaseLocales = databaseCharsets.get(database);
-            if (defaultDatabaseLocales != null) {
-                defaultDatabaseLocales.appendToDdlStatement(database, createDatabaseDddl);
-            }
-            addSchemaEvent(snapshotContext, database, createDatabaseDddl.toString());
-            addSchemaEvent(snapshotContext, database, "USE " + quote(database));
+                LOGGER.info("Reading structure of database '{}'", database);
+                addSchemaEvent(snapshotContext, database, "DROP DATABASE IF EXISTS " + quote(database));
+                final StringBuilder createDatabaseDdl = new StringBuilder("CREATE DATABASE " + quote(database));
+                final DatabaseLocales defaultDatabaseLocales = databaseCharsets.get(database);
+                if (defaultDatabaseLocales != null) {
+                    defaultDatabaseLocales.appendToDdlStatement(database, createDatabaseDdl);
+                }
+                addSchemaEvent(snapshotContext, database, createDatabaseDdl.toString());
+                addSchemaEvent(snapshotContext, database, "USE " + quote(database));
 
-            createSchemaEventsForTables(snapshotContext, tablesToRead.get(database), true);
+                createSchemaEventsForTables(snapshotContext, tablesToRead.get(database), true, executorService);
+            }
+        }
+        finally {
+            executorService.shutdownNow();
         }
     }
 
     void createSchemaEventsForTables(RelationalSnapshotContext<MySqlPartition, MySqlOffsetContext> snapshotContext,
-                                     final Collection<TableId> tablesToRead, final boolean firstPhase)
-            throws SQLException {
-        for (TableId tableId : tablesToRead) {
-            if (firstPhase && delayedSchemaSnapshotTables.contains(tableId)) {
-                continue;
+                                     final Collection<TableId> tablesToRead,
+                                     final boolean firstPhase,
+                                     ExecutorService executorService)
+            throws Exception {
+        CompletionService<Map<TableId, String>> completionService = new ExecutorCompletionService<>(executorService);
+        List<TableId> realTablesToRead;
+        if (firstPhase) {
+            realTablesToRead = tablesToRead.stream()
+                    .filter(id -> !delayedSchemaSnapshotTables.contains(id))
+                    .collect(Collectors.toList());
+        }
+        else {
+            realTablesToRead = new ArrayList<>(tablesToRead);
```

Comment:
Maybe you can simply this removing the else and move initialization to line 383.